### ISSUE TITLE
[X]-Science [v6.0.1.3], not picking up the science from the following biomes;

### ISSUE THREAD

**Issue by Vphurple-Izumi** [linuxgurugamer/KSP-X-Science#7]
Planet Kerbin;-
1. Woomerang Launch Site
2. Dessert Launch Site
3. Dessert Airfield

**Comment by Darael** [comment]
Correction: these are picked up, but not _correctly._

The checklist shows them as "Woomerang Launch Site (Shores)", "Dessert Launch Site (Shores)", and "Dessert Airfield (Shores)". The actual biomes as listed in [X] Science's "current location" display are "Woomerang Launch Site (Grasslands)", "Dessert Launch Site (Deserts)", and "Dessert Airfield (Deserts)". This has some… interesting consequences.

When in one of those biomes (let's look at the Dessert Launch Site, as an example), [X] Science Here & Now will list the current biome as "Dessert Launch Site (Deserts)", and will list no available experiments to run, even if there are in fact experiments that have not been run at the Dessert Launch Site. However, if you look at the main checklist window, experiments listed under the (shores) variant will acquire completion as they are run in the (deserts) variant.

I haven't yet tested, but I would suspect the same issue applies to any other "minibiomes" which are not contained in the Shores macrobiome. It is therefore worth checking the 5/6 "discoverable" launch site biomes, too (Mahi Mahi, Cove, Crater, Glacier Lake, and two at Baikerbanur) to see if they are also affected. The Island Airfield happens to be in the Shores biome, which is presumably why it is not impacted.

**Comment by Darael** [comment]
Updates:
Tested those discoverable launchsites. CoveLaunchSite's true parent biome is Grasslands (and it appears to be affected by this bug, showing current situation as "Landed at Kerbin's GlacierLakeLaunchSite (Tundra)" with no available experiments even though none have been run). CraterLaunchSite's is Shores, and as I suspected, it shows in Here and Now as it should. GlacierLakeLaunchSite is in Tundra, and is affected. MahiMahiLaunchSite is in Shores, and is not.

Baikerbanur is a little different — both Baikerbanur biomes have a parent of Grasslands (that is, "Baikerbanur (Grasslands)" and "Baikerbanur LaunchPad (Grasslands)"), but they aren't listed at all in the [X] Science checklist, whether it's been discovered yet or not, and accordingly the [X] Science Here and Now list is always empty when in those biomes, regardless of what experiments have or have not been run at them. This has the advantage of preserving the easter egg, but renders the Here and Now window _and_ the science checklist useless for those biomes.

So in short:
`CoveLaunchSite (Shores)`→`CoveLaunchSite (Grasslands)`
`GlacierLakeLaunchSite (Shores)`→`GlacierLakeLaunchSite (Tundra)`
`WoomerangLaunchSite (Shores)`→`WoomerangLaunchSite (Grasslands)`
`Dessert Launch Site (Shores)`→`Dessert Launch Site (Deserts)`
`Dessert Airfield (Shores)`→`Dessert Airfield (Deserts)`

…should resolve this bug, and #9 will require making some choices. I'm gonna see if I can do a PR for this myself.

**Comment by Darael** [comment]
Okay, my C# might _just_ be up to doing this myself but my knowledge of KSP internals is not.

However, I have identified the problem! Every collider-based biome is gathered in ScienceContext.cs, specifically, in `UpdateKscBiomes()`. That builds a simple list of every biome attached to a collider on Kerbin, *assuming that it is associated with KSC's biome*. This does not hold now that there are minibiomes in other macrobiomes.

It looks like there's going to need to be an appreciable amount of work to fix this; internal structures are using a simple list that assumes all minibiomes share a parent, which — since the decision was made to track the parent biome — would need to be changed to a list of pairs.

**Comment by SoullessCreature** [comment]
@Darael Have you developed a fix for this, or shall I have a crack at it?
